### PR TITLE
parameter dispather.

### PR DESCRIPTION
**Pull request by gongweibao**
Fix https://github.com/PaddlePaddle/Paddle/issues/8724
Fix https://github.com/PaddlePaddle/Paddle/issues/12488

### CODE REVIEW

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 1]
```diff
@@ -328,8 +333,93 @@ def get_trainer_program(self):
...
+            attrs={
+                "endpoints": self.pserver_endpoints,
+                RPC_OP_ROLE_ATTR_NAME: RPC_OP_ROLE_ATTR_VALUE
+            })
+
+        for varname, splited_var in self.param_var_mapping.iteritems():
+            # append concat_op to concat all splited params rather than prepend
+            if len(splited_var) <= 1:
+                continue
+            #splited_var_in_startup_prog = splited_param_mapping[varname]
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: remove comments

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 2]
```diff
@@ -328,8 +333,93 @@ def get_trainer_program(self):
...
+        """
+        if program is None:
+            program = self.startup_program
+
+        orig_s_prog = program
+
+        #delete initialize operators.
+        assert (orig_s_prog.num_blocks == 1)
+        # FIXME(gongwb): delete not need ops.
+        # note that: some parameter is not trainable and they ops can't be deleted.
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: they => these

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 3]
```diff
@@ -328,8 +333,93 @@ def get_trainer_program(self):
...
+        # delete_ops(orig_s_prog.global_block(), orig_s_prog.global_block().ops)
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: remove comment

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 4]
```diff
@@ -328,8 +333,93 @@ def get_trainer_program(self):
...
+
+        # add concat ops to origin parameters in startup program to
+        # let the origin parameters initialized by the spilited parameters
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: add concat ops to merge splited parameters received from parameter servers.

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 5]
```diff
@@ -328,8 +333,93 @@ def get_trainer_program(self):
...
         delete_ops(self.origin_program.global_block(), self.optimize_ops)
         self.origin_program.__str__()
+
         return self.origin_program
 
+    def _get_trainer_startup_program(self, program=None):
+        """
+        Get transpiled trainer side startup program.
+
+        Returns:
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: Should add Args

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 6]
```diff
@@ -328,8 +333,93 @@ def get_trainer_program(self):
...
+            Program: trainer side startup program.
+        """
+        if program is None:
+            program = self.startup_program
+
+        orig_s_prog = program
+
+        #delete initialize operators.
+        assert (orig_s_prog.num_blocks == 1)
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: Why need this assert, we currently have no places will have multiple blocks in the startup program

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 1 above)
Comment (gongweibao) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: Done.

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 2 above)
Comment (gongweibao) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: Done.

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 3 above)
Comment (gongweibao) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: Done.

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 4 above)
Comment (gongweibao) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: Done.

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 6 above)
Comment (gongweibao) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: Done.

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 5 above)
Comment (gongweibao) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: Done.

**Review by typhoonzero — APPROVED** [review]
LGTM++

**Pull request merged**